lib/src/Uncertainty/Algorithm/Simulation/PhysicalSpaceCrossEntropyImportanceSampling.cxx

```diff
@@ -0,0 +1,228 @@
+{
+  return new PhysicalSpaceCrossEntropyImportanceSampling(*this);
+}
+
+namespace {
+class KullbackLeiblerDivergenceObjective : public EvaluationImplementation
+{
+public:
+  KullbackLeiblerDivergenceObjective(const Sample auxiliaryCriticInputSample,
+                                     const Point & initialCriticInputSamplePDFValue,
+                                     const Distribution & auxiliaryDistribution,
+                                     const Indices & activeParameters,
+                                     const UnsignedInteger numberSamples)
+  : EvaluationImplementation()
+  , auxiliaryCriticInputSample_(auxiliaryCriticInputSample)
+  , activeParameters_(activeParameters)
+  , auxiliaryDistribution_(auxiliaryDistribution)
+  , initialCriticInputSamplePDFValue_(initialCriticInputSamplePDFValue)
+  , numberSamples_(numberSamples)
+  {
+    // Nothing to do
+  }
+  
+  KullbackLeiblerDivergenceObjective * clone () const override
+  {
+    return new KullbackLeiblerDivergenceObjective(*this);
+  }
+  
+  Point operator()(const Point & x) const override
+  {
+    Distribution distrib = auxiliaryDistribution_;
+  
+    // update auxiliary distribution
+    Point parameters(auxiliaryDistribution_.getParameter());
+    for (UnsignedInteger i = 0; i < activeParameters_.getSize(); ++i)
+    {
+      parameters[activeParameters_[i]] = x[i];
+    }
+    distrib.setParameter(parameters);
+  
+    const Point criticSamplesAuxiliaryPDFValue = distrib.computePDF(auxiliaryCriticInputSample_).asPoint();
+    const Point criticSamplesAuxiliaryLogPDFValue = distrib.computeLogPDF(auxiliaryCriticInputSample_).asPoint();
+  
+    Scalar objectiveFunction = 0.0;
+    for(UnsignedInteger j = 0; j < auxiliaryCriticInputSample_.getSize(); ++j)
+    {  
+      objectiveFunction += initialCriticInputSamplePDFValue_[j] / criticSamplesAuxiliaryPDFValue[j] * criticSamplesAuxiliaryLogPDFValue[j];
+    }
+    objectiveFunction = 1/numberSamples_*objectiveFunction;
+  
+    return Point(1,objectiveFunction);
+  }
+  
+  UnsignedInteger getInputDimension() const override
+  {
+    return activeParameters_.getSize();
+  }
+  
+  UnsignedInteger getOutputDimension() const override
+  {
+    return 1;
+  }
+  
+protected :
+  Sample auxiliaryCriticInputSample_;
+  Indices activeParameters_;
+  Distribution auxiliaryDistribution_;
+  Point initialCriticInputSamplePDFValue_;
+  Scalar numberSamples_;
+};
+}
+
+/** Set solver */
+void PhysicalSpaceCrossEntropyImportanceSampling::setOptimizationAlgorithm(const OptimizationAlgorithm & solver)
+{
+   solver_= solver;
+}
+
+/** Get solver */
+OptimizationAlgorithm PhysicalSpaceCrossEntropyImportanceSampling::getOptimizationAlgorithm() const
+{
+  return solver_;
+}
+
+
+// Compute Output Samples
+Sample PhysicalSpaceCrossEntropyImportanceSampling::computeOutputSamples(const Sample & inputSamples) const
+{
+  Sample outputSamples = getEvent().getFunction()(inputSamples);
```

Comment:
As `Sample` is a complex type, and as it will not be modified afterward, you should write:
```
const Sample outputSamples(getEvent().getFunction()(inputSamples));
```
but there is no need to introduce a temporary variable. Simply write:
```
return getEvent().getFunction()(inputSamples);
```

Reply:
This is now fixed.